init: Take lock on blocks directory in BlockManager ctor

sedited opened the pull request:
This ensures consistent directory locking for users of the kernel library by disallowing mistakenly creating multiple BlockManagers that might write into the same block or undo files.

The change here makes `LockDirectory` more strict: It now returns an error even if it is called again on the same directory from the same process. However from the description in https://github.com/bitcoin/bitcoin/pull/12422 , where this feature was introduced, it is not immediately clear what its usage was at the time. Supporting multiple wallets in the same directory is mentioned there, but it is not said why that might be a desirable feature.

sedited committed:
util: Prevent multiple LockDirectory calls within the same process

Previously LockDirectory only prevented concurrent locks across
different processes, but allowed the same process to re-lock on the same
directory.

This change is not immediately relevant for its current use, where the
lock is only supposed to protect against a different process writing on
the same resources.

This change is relevant for future use by the kernel library, where
users of the library might mistakenly create multiple instances of an
object that seek to write to a common resource.

stickies-v reviewed (commented):
Concept ACK on:
- making LockDirectory more strict, it confused me recently that I was able to acquire a lock multiple times (within the same process)
- making BlockManager manage the lock instead of init

As per https://github.com/TheCharlatan/bitcoin/pull/25, I would be a fan of going further and making the `BlockManagerOptions` ctor take a `BlockDirLock` (or, as per my comment a `DirectoryLock`) to allow more granular error feedback, but this is already a step in the right direction regardless.

In the future, this approach would also make it more straightforward to e.g. let `BlockManager` give other threads read-only access to the blocksdir via something like:

```
class DirectoryLock {
private:
    std::shared_mutex m_mutex;
public:
    const fs::path m_path;

    DirectoryLock(const fs::path& path)
        : m_path(path)
    {
        assert(util::LockDirectory(m_path, ".lock", false) == util::LockResult::Success);
    }
    ~DirectoryLock() { UnlockDirectory(m_path, ".lock"); }

    DirectoryLock(const DirectoryLock&) = delete;
    DirectoryLock& operator=(const DirectoryLock&) = delete;

    std::unique_lock<std::shared_mutex> lock_exclusive()
    {
        return std::unique_lock<std::shared_mutex>(m_mutex);
    }
    std::shared_lock<std::shared_mutex> lock_shared()
    {
        return std::shared_lock<std::shared_mutex>(m_mutex);
    }
};
```

stickies-v commented on src/node/blockstorage.h:
```diff
@@ -127,6 +127,14 @@ struct BlockfileCursor {
+{
+    fs::path m_path;
+
+public:
```
And have `BlockManager::m_lock` be a `std::unique_ptr<BlockDirLock>` instead?

TheCharlatan replied:
I've been thinking about allocating it on the heap instead, but I kind of like that declaring it as the first field forces us to initialize it first, which a pointer type would not necessarily. Then again, moving the lock as a unique pointer from the options to the actual class would also probably be easier than defining a move constructor the directory lock class.

sedited committed:
util: Add RAII directory lock

This makes it easier for a class or a struct to own a lock on a
directory for the duration of its lifetime. It is used in the next
commit.

sedited committed:
init: Take lock on blocks directory in BlockManager ctor

This moves the responsibility of taking the lock for the blocks
directory into the BlockManager. Use the DirectoryLock wrapper to ensure
it is the first resource to be acquired and is released again after use.

This is relevant for the kernel library where the lock should be taken
even if the user fails to explicitly do so.

TheCharlatan commented:
> I would be a fan of going further and making the BlockManagerOptions ctor take a BlockDirLock (or, as per my comment a DirectoryLock) to allow more granular error feedback, but this is already a step in the right direction regardless.

I attempted implementing this, but giving the options the lock starts to introduce non-trivial additional semantics to the options: Since the lock needs to be moved from the options to the block manager, the options cannot be re-used. I'm not sure how to implement that and make it clear to the user what is going on.
